spec/services/form_validator_spec.rb

```diff
@@ -0,0 +1,198 @@
+require 'rails_helper'
+
+describe FormValidator do
+
+  let(:form) { create :form }
+  let(:element) { create :form_element, form: form, label: 'test' }
+
+  before :each do
+    @params = {form_id: element.form_id}
+  end
+
+  describe "errors and save" do
+
+    describe "adds an error if" do
+
+      after :each do
+        expect(@validator.valid?).to eq false
+      end
+      
+      describe "required field" do
+
+        before :each do
+          element.update_attributes({required: true})
+        end
+
+        after :each do
+          @validator = FormValidator.new(@params)
+          expect(@validator.errors).to eq ({test: ['is required']})
+        end
+
+        it "does not appear" do
+          @params.merge!({})
+        end
+
+        it "is nil" do
+          @params.merge!(test: nil)
+        end
+
+        it "is false" do
+          @params.merge!(test: false)
+        end
+
+        it "is empty string" do
+          @params.merge!(test: "")
+        end
+      end
+
+      describe "email field" do
+
+        before :each do
+          element.update_attributes({data_type: "email"})
+        end
+
+        after :each do
+          @validator = FormValidator.new(@params)
+          expect(@validator.errors).to eq ({test: [I18n.t("validation.is_invalid_email")]})
+        end
+
+        it "is a basic sentence" do
+          @params.merge!(test: "I'm not an email!")
+        end
+
+        it "is missing the TLD" do
+          @params.merge!(test: "neal@sumofus")
+        end
+
+        it "has two @ symbols" do
+          @params.merge!(test: "this@that@other.com")
+        end
+      end
+
+      describe "phone" do
+
+        before :each do
+          element.update_attributes({data_type: "phone"})
+        end
+
+        after :each do
+          @validator = FormValidator.new(@params)
+          expect(@validator.errors).to eq ({test: [I18n.t("validation.is_invalid_phone")]})
```

Comment:
or do a pattern match? `r%{have numbers, dash}`... since I'm not sure what rails displays when the translation is missing - it might be an error message that is a string of a decent length.